src/util/inc/tcache.h

```diff
@@ -54,6 +54,7 @@ typedef struct SCacheDataNode {
   uint32_t           size;         // allocated size for current SCacheDataNode
   T_REF_DECLARE()
   char              *key;
+  unsigned int       data_freed:2;
```

Comment:
I am not quit clear why "data_freed" is needed?

Reply:
it's been a while since last PR'd.

please be kindly noted:
```
137   if (pCacheObj->freeFp) {
138     DASSERT(pElem->pData->data_freed==0);
139     if (!pElem->pData->data_freed) {
140       pCacheObj->freeFp(pElem->pData->data);
141       pElem->pData->data_freed = 1;
142     }
143   }
```
DASSERT is there to help find `double-free` problems.

well, it exists because cached object could be freed within cacher-timer-thread or user-thread if we can say so, we
need a way to prohibit freeFp from being called more than once. from the critical point of view, `data-freed` is not as good as expected. 


this PR's main purpose was to re-implement a better hash. modifications in tcache.c, including data-freed, is not
a firm and solid one. and i am still whatsoever confused by the cache api semantics brought in tache.h. because tcache's been used heavenly in the project, i hesitated to modify tcache at the very beginning.

wish this helps.